collectives/utils/mail.py

```diff
@@ -75,4 +76,4 @@ def send_mail(**kwargs):
         )
         msg["DKIM-Signature"] = sig.decode("ascii").lstrip("DKIM-Signature: ")
 
-    s.send_message(msg)
+    threading.Thread(target=s.send_message, args=(msg,)).start()
```

Comment:
SI je lis bien la documentation, j'ai l'impression que toute exception levée dans le thread sera ignorée (enfin, juste logguée sur stderr). Il faudrait capturer les erreurs d'envoi d'email pour en informer l'utilisateur (qu'il n'attende pas l'email de confirmation pour rien)

Edit: cf https://stackoverflow.com/questions/2829329/catch-a-threads-exception-in-the-caller-thread-in-python

Reply:
Hum, je comprends ce que tu veux dire. Certaine erreur de mail peuvent ne pas être retournée à l'utilisateur. EG: un utilisateur se désinscrivant n'est pas intéressé de savoir que le mail à l'encadrant est mal parti. Cependant, c'est un peu plus gènant pour les mails de confirmation. J'ai rajouté un système de gestion des erreurs dans le thread. Dans notre cas, ca veut que l'utilisateur est prévenu de l'échec dès qu'il recharge une page /auth . Est-ce que ca te va comme ça?

Reply:
Oui, bonne remarque. Mais dans ce cas, ne serait-il pas possible d'utiliser la version threadé pour les envois de notifications (desinscription, etc), et d'utiliser la version synchrone pour l'inscription/récupération ? Ces actions sont faites assez rarement, et je ne pense pas que les utilisateurs soient choqués de devoir attendre une seconde pour cela. J'ai peur que les utilisateurs ne pensent pas à rafraichir la page pour avoir le message d'erreur

Reply:
>  je ne pense pas que les utilisateurs soient choqués de devoir attendre une seconde pour cela. 

En cas d'échec, ca peut prendre 5-10-15 secondes, et on risque d'avoir des personnes qui rafraichissent, des timeout en 502, etc... Je suis vraiment partisan de ne pas mettre d'actions longues dans les threads de génération de page.

Plutôt de montrer l'erreur au rechargement d'une page, je ferais une action AJAX alors.

A noter qu'il serait intéressant de mettre aussi les requètes à l'extranet dans des threads séparés: en cas de problème de connexion (DNS lents, coupure réseau, etc...) on peut perdre énormément de temps. Mais ça sera pour une autre fois

Reply:
Pour l'extranet c'est plus compliqué, dans la plupart des cas on a besoin de la réponse (savoir si la licence est à jour) pour rendre la page. Du coup, une IO asynchrone n'apporterait pas grand chose (le thread flask aurait besoin d'attendre le thread extranet, et contrairement à d'autre paradigmes genre node.js il ne serait pas récupéré par flask pour répondre à une autre requête en attendant). 
Si on veut être sur que la requête ne s'éternise pas il est surement possible de définir un timeout sur la urllib python

Reply:
Après il y a d'autres taches qui gagneraient à être faites en arrière plan, par ex le redimensionnement d'avatars, le log des statistiques, etc, tout ce dont on a pas besoin du résultat immédiatement. Dans ce cas un truc genre Celery pourrait être pas mal

Reply:
> Après il y a d'autres taches qui gagneraient à être faites en arrière plan, par ex le redimensionnement d'avatars, le log des statistiques, etc, tout ce dont on a pas besoin du résultat immédiatement.

Je suis d'accord avec toi. Je me suis concentré sur le mail, mais il y a d'autres sujets qu'on pourrait traiter

Reply:
> contrairement à d'autre paradigmes genre node.js il ne serait pas récupéré par flask pour répondre à une autre requête en attendant

C'est vrai que flask est pas le plus indiqué pour ça, mais on devrait pouvoir y arriver. C'est juste que c'est toujours frustrant de voir une page qui met du temps à se charger!

Reply:
@gdaviet J'ai fait un système qui gère le statut du token. Ca permet à l'utilisateur de voir quand il est en train d'être envoyé, et si il a été envoyé avec succès ou non. Je te laisse review ça, mais à moins qu'il y ai un gros bug, j'aimerais vraiment qu'on puisse merger ça: le but est pas forcément d'avoir un système parfait, mais au moins d'avancer dans la bonne direction.